src/WebJobs.Extensions.DurableTask/EntityScheduler/MessageSorter.cs

```diff
@@ -0,0 +1,243 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    /// <summary>
+    /// provides message ordering and deduplication of request messages (operations or lock requests)
+    /// that are sent to entities, from other entities, or from orchestrations.
+    /// </summary>
+    [JsonObject(MemberSerialization.OptIn)]
+    internal class MessageSorter
+    {
+        // don't update the reorder window too often since the garbage collection incurs some overhead.
+        private static readonly TimeSpan MinIntervalBetweenCollections = TimeSpan.FromSeconds(10);
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, DateTime> Sent { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public Dictionary<string, ReceiveBuffer> Received { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime ReceiveHorizon { get; set; }
+
+        [JsonProperty(DefaultValueHandling = DefaultValueHandling.Ignore)]
+        public DateTime SendHorizon { get; set; }
+
+        public int NumberBufferedRequests =>
+            this.Received?.Select(kvp => kvp.Value.Buffered?.Count ?? 0).Sum() ?? 0;
+
+        /// <summary>
+        /// Called on the sending side, to fill in timestamp and predecessor fields.
+        /// </summary>
+        public void LabelOutgoingMessage(RequestMessage message, string destination, DateTime now, TimeSpan reorderWindow)
+        {
+            DateTime timestamp = now;
+
+            if (this.SendHorizon + reorderWindow + MinIntervalBetweenCollections < now)
+            {
+                this.SendHorizon = now - reorderWindow;
+
+                // clean out send clocks that are past the reorder window
+
+                if (this.Sent != null)
+                {
+                    List<string> expired = null;
+
+                    foreach (var kvp in this.Sent)
+                    {
+                        if (kvp.Value < this.SendHorizon)
+                        {
+                            (expired ?? (expired = new List<string>())).Add(kvp.Key);
+                        }
+                    }
+
+                    if (expired != null)
+                    {
+                        foreach (var t in expired)
+                        {
+                            this.Sent.Remove(t);
+                        }
+
+                        expired.Clear();
+                    }
+
+                    if (this.Sent.Count == 0)
+                    {
+                        this.Sent = null;
+                    }
+                }
+            }
+
+            if (this.Sent == null)
+            {
+                this.Sent = new Dictionary<string, DateTime>();
```

Comment:
Should these dictionaries (`Sent` and `Received`) use case-insensitive key lookups?

Normally whenever I see `Dictionary<string, TValue>` I ask myself whether this dictionary needs to use case-insensitive comparisons. I haven't reviewed closely enough to know whether `destination` could potentially be specified in different casings.

Reply:
Good question. These strings are instance ids. It looks to me like you are using `StringComparer.OrdinalIgnoreCase` for instance ids (at least in DurableTask.AzureStorage.OrchestrationSessionManager.cs) so I will use the same here.